File: src/main/scala/com/exasol/cloudetl/bucket/AzureAbfsBucket.scala
package com.exasol.cloudetl.bucket

import scala.util.matching.Regex

import com.exasol.cloudetl.storage.StorageProperties
import com.exasol.errorreporting.ExaError

import org.apache.hadoop.conf.Configuration

/** A [[Bucket]] implementation for the Azure Data Lake Gen2 Storage */
final case class AzureAbfsBucket(path: String, params: StorageProperties) extends Bucket with SecureBucket {

  private[this] val AZURE_ACCOUNT_NAME: String = "AZURE_ACCOUNT_NAME"
  private[this] val AZURE_SECRET_KEY: String = "AZURE_SECRET_KEY"

  /** @inheritdoc */
  override val bucketPath: String = path

  /** @inheritdoc */
  override val properties: StorageProperties = params

  /**
   * Returns the list of required property keys for Azure Data Lake Gen2
   * Storage.
   */
  override def getRequiredProperties(): Seq[String] =
    Seq.empty[String]

  /** @inheritdoc */
  override def getSecureProperties(): Seq[String] =
    Seq(AZURE_SECRET_KEY)

  /** @inheritdoc */
  override def validate(): Unit = {
    validateRequiredProperties()
    validateConnectionProperties()
  }

  /** @inheritdoc */
  override def getConfiguration(): Configuration = {
    validate()

    val conf = new Configuration()
    conf.set("fs.abfs.impl", classOf[org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem].getName())
    conf.set("fs.AbstractFileSystem.abfs.impl", classOf[org.apache.hadoop.fs.azurebfs.Abfs].getName())
    conf.set("fs.abfss.impl", classOf[org.apache.hadoop.fs.azurebfs.SecureAzureBlobFileSystem].getName())
    conf.set("fs.AbstractFileSystem.abfss.impl", classOf[org.apache.hadoop.fs.azurebfs.Abfss].getName())

    val mergedProperties = if (properties.hasNamedConnection()) {
      properties.merge(AZURE_ACCOUNT_NAME)
    } else {
      properties
    }

    val accountAndContainer = regexParsePath(path)
    val accountName = mergedProperties
      .get(AZURE_ACCOUNT_NAME)
      .getOrElse(accountAndContainer.accountName)
    val secretKey = mergedProperties.getString(AZURE_SECRET_KEY)

    conf.set(s"fs.azure.account.key.$accountName.dfs.core.windows.net", secretKey)

    conf
  }

  // Intentionally copy-paste, duplicate count: 2. Please, refactor when
  // it reaches 3+.
  // Fabric / OneLake: .dfs.fabric.microsoft.com
  private[this] final val AZURE_ABFS_PATH_REGEX: Regex =
    """abfss?://(.*)@([^.]+).dfs.core.windows.net/(.*)$""".r
  private[this] final val AZURE_ABFS_ONELAKE_PATH_REGEX: Regex =
    """abfss?://(.*)@([^.]+)\.dfs\.fabric\.microsoft\.com/(.*)$""".r

  private[this] def regexParsePath(path: String): AccountAndContainer = path match {
    case AZURE_ABFS_PATH_REGEX(containerName, accountName, _) =>
      AccountAndContainer(accountName, containerName)
    case AZURE_ABFS_ONELAKE_PATH_REGEX(containerName, accountName, _) =>
      AccountAndContainer(accountName, containerName)
    case _ =>
      throw new BucketValidationException(
        ExaError
          .messageBuilder("E-CSE-20")
          .message("Azure datalake storage path {{PATH}} scheme is not valid.", path)
          .mitigation("It should be either 'abfs' or 'abfss'.")
          .toString()
      )
  }

  private[this] case class AccountAndContainer(accountName: String, containerName: String)

}

File: src/main/java/com/exasol/cloudetl/bucket/AzureAbfsBucket.java
package com.exasol.cloudetl.bucket;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.hadoop.conf.Configuration;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.errorreporting.ExaError;

/** Bucket implementation for Azure Data Lake Gen2 Storage. */
public final class AzureAbfsBucket extends AbstractConfiguredBucket implements SecureBucket {
    private static final String AZURE_ACCOUNT_NAME = "AZURE_ACCOUNT_NAME";
    private static final String AZURE_SECRET_KEY = "AZURE_SECRET_KEY";
    private static final Pattern AZURE_ABFS_PATH_REGEX = Pattern
            .compile("abfss?://(.*)@([^.]+).dfs.core.windows.net/(.*)$");

    /** Create a bucket. */
    public AzureAbfsBucket(final String path, final StorageProperties params) {
        super(path, params);
    }

    @Override
    public boolean equals(final Object obj) {
        return (this == obj) || ((obj instanceof AzureAbfsBucket) && hasSameConfiguration((AzureAbfsBucket) obj));
    }

    @Override
    public int hashCode() {
        return configuredHashCode(AzureAbfsBucket.class);
    }

    @Override
    public scala.collection.immutable.Seq<String> getRequiredProperties() {
        return ScalaConverters.seqFromJava(List.of());
    }

    @Override
    public scala.collection.immutable.Seq<String> getSecureProperties() {
        return ScalaConverters.seqFromJava(List.of(AZURE_SECRET_KEY));
    }

    @Override
    public void validate() {
        validateRequiredProperties();
        validateConnectionProperties();
    }

    @Override
    public Configuration getConfiguration() {
        validate();
        final Configuration conf = new Configuration();
        conf.set("fs.abfs.impl", org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.class.getName());
        conf.set("fs.AbstractFileSystem.abfs.impl", org.apache.hadoop.fs.azurebfs.Abfs.class.getName());
        conf.set("fs.abfss.impl", org.apache.hadoop.fs.azurebfs.SecureAzureBlobFileSystem.class.getName());
        conf.set("fs.AbstractFileSystem.abfss.impl", org.apache.hadoop.fs.azurebfs.Abfss.class.getName());
        final StorageProperties mergedProperties = properties().hasNamedConnection() ? properties().merge(AZURE_ACCOUNT_NAME)
                : properties();
        final AccountAndContainer accountAndContainer = regexParsePath(bucketPath());
        final String accountName = mergedProperties.get(AZURE_ACCOUNT_NAME).isDefined()
                ? mergedProperties.get(AZURE_ACCOUNT_NAME).get()
                : accountAndContainer.accountName;
        conf.set("fs.azure.account.key." + accountName + ".dfs.core.windows.net",
                mergedProperties.getString(AZURE_SECRET_KEY));
        return conf;
    }

    private AccountAndContainer regexParsePath(final String path) {
        final Matcher matcher = AZURE_ABFS_PATH_REGEX.matcher(path);
        if (matcher.matches()) {
            return new AccountAndContainer(matcher.group(2), matcher.group(1));
        }
        throw new BucketValidationException(ExaError.messageBuilder("E-CSE-20")
                .message("Azure datalake storage path {{PATH}} scheme is not valid.", path)
                .mitigation("It should be either 'abfs' or 'abfss'.").toString());
    }

    private static final class AccountAndContainer {
        private final String accountName;

        private AccountAndContainer(final String accountName, final String containerName) {
            this.accountName = accountName;
        }
    }
}
